modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleService.java

```diff
@@ -582,7 +568,7 @@ private Set<Index> evaluateDownsampleStatus(
      */
     private void replaceBackingIndexWithDownsampleIndexOnce(DataStream dataStream, String backingIndexName, String downsampleIndexName) {
         clusterStateChangesDeduplicator.executeOnce(
-            new ReplaceSourceWithDownsampleIndexTask(dataStream.getName(), backingIndexName, downsampleIndexName, null),
+            new DeleteSourceAndAddDownsampleToDS(settings, dataStream.getName(), backingIndexName, downsampleIndexName, null),
```

Comment:
This usage really threw me off at first. I was confused why we were constructing two DeleteSourceAndAddDownsampleToDS objects here. The answer is that the first one is used as a dedup key because DeleteSourceAndAddDownsampleToDS's equals/hashCode don't take its settings or listener into account. The second one is the one that is actually used to delete/add the indices, and the first one is just thrown away.
I see why you did it, and it technically works. It just seems confusing and maybe violates the expectations of equals/hashCode, as well as the expectations of executeOnce.

Reply:
Yeah, we have a test to make sure the equals/hashcode obeys the invariants we want it to obey for the dedup reasons.

Reply:
Would it be clearer to just pass a String here -- `dataStream.getName() + "-" + backingIndexName + "-" + downsampleIndexName` to make it clear that all we're using this for is dedup? Or add a comment "This object is used purely as a dedup key and not as a ClusterStateTaskListener. This object is dropped, and a new ClusterStateTaskListener is created below." I think it could avoid some confusion in the future.

Reply:
Oh, I think I like using the `String` type for dedup idea ... very much !